Modify rule S4275: Adapt to LaYC (#2105)
SonarSource/rspec: rules/S4275/javascript/rule.adoc

== Why is this an issue?
Getters and setters provide a way to enforce encapsulation by providing methods that give controlled access to class fields. However, in classes with multiple fields, it is not unusual that copy and paste is used to quickly create the needed getters and setters, which can result in the wrong field being accessed by a getter or setter.

> This rule raises an issue in any of these cases:

Reviewer: ```suggestion
This rule raises an issue in the following cases.
```

* A setter does not update the field with the corresponding name (if it exists).

> * A getter does not return any value or does not access the field with the corresponding name (if it exists).

Reviewer: Let's add code example where getter does not return anything (may be even conditionally)

> * A getter does not return any value or does not access the field with the corresponding name (if it exists).

Reviewer: Did you consider splitting this point into two? 
* getter does not return anything
* getter returns the wrong thing

Underscore prefixes for fields are supported, so `setX()` can assign a value to `_x`.

The following type of getters and setters are supported:

* `getX()` and `setX()`

[source,javascript,diff-id=1,diff-type=noncompliant]
----
class A {
  #y: number = 0;
  setY(val: number) { // Noncompliant: field '#y' is not updated
  }
}
----
[source,javascript,diff-id=1,diff-type=noncompliant]
----
class A {
  #y: number = 0;
  setY(val: number) {
    this.#y = val;
  }
}
----
* `get x()` and `set x()`

[source,javascript,diff-id=2,diff-type=noncompliant]
----
class A {
  x: number = 0;
  #y: number = 0;

  get x() { // Noncompliant: field 'x' is not used in the return value
    return this.#y;
  }
}
----
[source,javascript,diff-id=2,diff-type=compliant]
----
class A {

>   _x: number = 0;

Reviewer: you renamed this variable compared to the previous code sample

  }
});
----

== Resources

> === Documentation

Reviewer: I suggest to add this page as well
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Private_class_fields

> - https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/get[MDN get]

Reviewer: Add `-` after MDN
